wpimath/src/main/java/edu/wpi/first/math/autodiff/VariableMatrix.java

```diff
@@ -0,0 +1,1029 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.math.autodiff;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.BinaryOperator;
+import java.util.function.UnaryOperator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.ejml.simple.SimpleMatrix;
+
+/** A matrix of autodiff variables. */
+public class VariableMatrix implements Iterable<Variable> {
+  private final ArrayList<Variable> m_storage = new ArrayList<>();
+  private int m_rows;
+  private int m_cols;
+
+  /** Constructs an empty VariableMatrix. */
+  public VariableMatrix() {}
+
+  /**
+   * Constructs a VariableMatrix from Variable internal handles.
+   *
+   * <p>This constructor is for internal use only.
+   *
+   * @param rows The number of matrix rows.
+   * @param cols The number of matrix columns.
+   * @param handles Variable handles.
+   */
+  public VariableMatrix(int rows, int cols, long[] handles) {
+    m_rows = rows;
+    m_cols = cols;
+    for (int index = 0; index < rows * cols; ++index) {
+      m_storage.add(new Variable(Variable.HANDLE, handles[index]));
+    }
+  }
```

Comment:
Should we check that `handles.length == rows * cols`? This is only intended for internal use, so we may not need to, but it would definitely be a weird situation if `size() != cols() * rows()`.